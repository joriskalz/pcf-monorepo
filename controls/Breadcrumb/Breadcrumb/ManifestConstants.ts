export const enum ManifestPropertyNames {
    dataset = 'dataset',
}

export const enum ItemColumns {
    DisplayName = 'ItemDisplayName',
    Key = 'ItemKey',
    Clickable = 'ItemClickable',
}

export const enum InputEvents {
    SetFocus = 'SetFocus',
}

export const enum InputProperties {
    InputEvent = 'InputEvent',
    SelectedKey = 'SelectedKey',
}
