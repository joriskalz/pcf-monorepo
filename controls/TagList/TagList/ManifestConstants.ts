export const enum ManifestPropertyNames {
    dataset = 'dataset',
}

export const enum ItemColumns {
    DisplayName = 'ItemDisplayName',
    Key = 'ItemKey',
    Enabled = 'ItemEnabled',
    IconName = 'ItemIconName',
    IconColor = 'ItemIconColor',
    TextColor = 'ItemTextColor',
    BackgroundColor = 'ItemBackgroundColor',
    BorderColor = 'ItemBorderColor',
    IconOnly = 'ItemIconOnly',
    Visible = 'ItemVisible',
}

export const TextAlignmentTypes = {
    '0': 'center',
    '1': 'left',
    '2': 'right',
};
