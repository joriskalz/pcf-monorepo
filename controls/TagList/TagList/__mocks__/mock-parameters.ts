/* istanbul ignore file */

import { IInputs } from '../generated/ManifestTypes';
import { MockEnumProperty, MockStringProperty, MockWholeNumberProperty } from './mock-context';
import { MockDataSet } from './mock-datasets';

export function getMockParameters(): IInputs {
    return {
        AccessibilityLabel: new MockStringProperty(),
        BorderRadius: new MockWholeNumberProperty(),
        SelectedKey: new MockStringProperty(),
        Theme: new MockStringProperty(),
        TextAlignment: new MockEnumProperty(),
        items: new MockDataSet([]),
        MaxHeight: new MockWholeNumberProperty(),
        FontSize: new MockWholeNumberProperty(),
        ItemHeight: new MockWholeNumberProperty(),
    };
}
