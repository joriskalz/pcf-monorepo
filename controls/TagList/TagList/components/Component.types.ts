export interface TagListItem {
    id: string;
    enabled?: boolean;
    name: string;
    iconName?: string;
    iconColor?: string;
    backgroundColor?: string;
    borderColor?: string;
    textColor?: string;
    iconOnly?: boolean;
    visible?: boolean;
    parentItemKey?: string;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    data: any;
    key: string;
}

export interface CanvasTagListProps {
    width?: number;
    height?: number;
    items: TagListItem[];
    disabled?: boolean;
    tabIndex?: number;
    ariaLabel?: string;
    themeJSON?: string;
    itemMaxWidth?: number;
    maxHeight?: number;
    itemHeight?: number;
    fontSize?: number;
    borderRadius?: number;
    justify?: 'center' | 'left' | 'right';
    onResize?: (width: number, height: number) => void;
}
