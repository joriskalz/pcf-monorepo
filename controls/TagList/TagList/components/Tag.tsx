import { useId } from '@fluentui/react-hooks';
import {
    FontIcon,
    classNamesFunction,
    styled,
    ITagItemProps,
    ITagItemStyleProps,
    ITagItemStyles,
} from '@fluentui/react';
import { getStyles } from './TagItem.styles';
import * as React from 'react';

const getClassNames = classNamesFunction<ITagItemStyleProps, ITagItemStyles>();
export interface TagProps extends ITagItemProps {
    fontSize?: number;
    height?: number;
    iconName?: string;
    iconColor?: string;
    textColor?: string;
    backgroundColor?: string;
    borderColor?: string;
    borderRadius?: number;
}

// This is simlar to the TagItem in Fluent
export const TagComponent = (props: TagProps): JSX.Element => {
    const {
        theme,
        styles,
        iconColor,
        textColor,
        backgroundColor,
        borderColor,
        iconName,
        children,
        index,
        title = typeof props.children === 'string' ? props.children : props.item.name,
    } = props;

    const tagStyle = React.useMemo(() => {
        return {
            color: textColor,
            backgroundColor: backgroundColor,
            borderColor: borderColor,
            borderWidth: borderColor ? 1 : undefined,
            borderStyle: borderColor ? 'solid' : undefined,
        } as React.CSSProperties;
    }, [textColor, backgroundColor, borderColor]);

    const iconStyle = React.useMemo(() => {
        return {
            color: iconColor,
        } as React.CSSProperties;
    }, [iconColor]);

    const itemId = useId();
    const classNames = getClassNames(styles, {
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        theme: theme!,
        disabled: true,
        selected: false,
    });

    return (
        <div className={classNames.root} style={tagStyle} role={'listitem'} aria-label={title} key={index}>
            {iconName && <FontIcon role="img" iconName={iconName} style={iconStyle} />}
            <span className={classNames.text} title={title} id={`${itemId}-text`}>
                {children}
            </span>
        </div>
    );
};

export const Tag = React.memo(
    styled<TagProps, ITagItemStyleProps, ITagItemStyles>(TagComponent, getStyles, undefined, {
        scope: 'TagItem',
    }),
);
