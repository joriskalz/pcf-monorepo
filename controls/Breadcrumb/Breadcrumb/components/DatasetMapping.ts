import { IBreadcrumbItem } from '@fluentui/react';
import { ItemColumns } from '../ManifestConstants';
import { ICustomBreadcrumbItem } from './components.types';

export function getItemsFromDataset(dataset: ComponentFramework.PropertyTypes.DataSet): ICustomBreadcrumbItem[] {
    if (dataset.error || dataset.paging.totalResultCount === undefined) {
        // Dataset is not defined so return dummy items
        return [getDummyAction('1'), getDummyAction('2'), getDummyAction('3')];
    }
    const keyIndex: Record<string, number> = {};
    return dataset.sortedRecordIds.map((id) => {
        const record = dataset.records[id];
        // Prevent duplicate keys by appending the duplicate index
        let key = record.getValue(ItemColumns.Key) as string;
        if (keyIndex[key] !== undefined) {
            keyIndex[key]++;
            key += `_${keyIndex[key]}`;
        } else keyIndex[key] = 1;
        return {
            id: record.getRecordId(),
            key: key,
            text: record.getFormattedValue(ItemColumns.DisplayName),
            clickable: record.getValue(ItemColumns.Clickable) as boolean,
            data: record,
        } as ICustomBreadcrumbItem;
    });
}

export function getBreadcrumbItems(
    items: ICustomBreadcrumbItem[],
    onClick: (ev?: React.MouseEvent<HTMLElement>, item?: ICustomBreadcrumbItem) => void,
): IBreadcrumbItem[] {
    return items.map((item) =>
        item.clickable === false
            ? {
                  id: item.id,
                  key: item.key,
                  text: item.text,
              }
            : {
                  id: item.id,
                  key: item.key,
                  text: item.text,
                  onClick: onClick,
              },
    ) as IBreadcrumbItem[];
}

function getDummyAction(key: string): ICustomBreadcrumbItem {
    return {
        id: key,
        key: key,
        text: 'Item ' + key,
        clickable: true,
    } as ICustomBreadcrumbItem;
}
