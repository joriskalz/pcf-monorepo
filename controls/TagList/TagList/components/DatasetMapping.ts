import { TagListItem } from './Component.types';
import { ItemColumns } from '../ManifestConstants';

export function getTagItemsFromDataset(dataset: ComponentFramework.PropertyTypes.DataSet): TagListItem[] {
    if (dataset.error || dataset.paging.totalResultCount === undefined) {
        // Dataset is not defined so return dummy items
        return [getDummyAction('1'), getDummyAction('2'), getDummyAction('3')];
    }
    const keyIndex: Record<string, number> = {};
    return dataset.sortedRecordIds.map((id) => {
        const record = dataset.records[id];
        // Prevent duplicate keys by appending the duplicate index
        let key = record.getValue(ItemColumns.Key) as string;
        if (keyIndex[key] !== undefined) {
            keyIndex[key]++;
            key += `_${keyIndex[key]}`;
        } else keyIndex[key] = 1;
        return {
            id: record.getRecordId(),
            key: key,
            name: record.getValue(ItemColumns.DisplayName) as string,
            enabled: record.getValue(ItemColumns.Enabled) as boolean,
            iconName: record.getValue(ItemColumns.IconName) as string,
            iconColor: record.getValue(ItemColumns.IconColor) as string,
            textColor: record.getValue(ItemColumns.TextColor) as string,
            borderColor: record.getValue(ItemColumns.BorderColor) as string,
            backgroundColor: record.getValue(ItemColumns.BackgroundColor) as string,
            iconOnly: record.getValue(ItemColumns.IconOnly) as boolean,
            visible: record.getValue(ItemColumns.Visible) as boolean,
            data: record,
        } as TagListItem;
    });
}

function getDummyAction(key: string): TagListItem {
    return {
        id: key,
        key: key,
        name: 'Item ' + key,
        iconName: 'Unknown',
    } as TagListItem;
}
