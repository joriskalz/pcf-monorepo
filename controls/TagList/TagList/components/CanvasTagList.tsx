import {
    ITagItemStyleProps,
    ITagItemStyles,
    ThemeProvider,
    IStyleFunctionOrObject,
    createTheme,
    IPartialTheme,
} from '@fluentui/react';
import * as React from 'react';
import { Tag } from './Tag';
import { TagListItem, CanvasTagListProps } from './Component.types';
import useResizeObserver from '@react-hook/resize-observer';

export const CanvasTagList = React.memo((props: CanvasTagListProps) => {
    const {
        items,
        width,
        height,
        maxHeight,
        itemMaxWidth,
        disabled,
        tabIndex,
        themeJSON,
        justify,
        onResize,
        itemHeight,
        fontSize,
        borderRadius,
    } = props;

    const containerStyle: React.CSSProperties = React.useMemo(() => {
        return {
            maxWidth: width ? width : '100%',
            position: 'relative',
            display: 'flex',
            maxHeight: maxHeight || height,
            overflowY: 'auto',
            flexWrap: 'wrap',
            alignContent: 'flex-start',
        };
    }, [height, maxHeight, width]);

    const tagStyles: IStyleFunctionOrObject<ITagItemStyleProps, ITagItemStyles> = React.useMemo(() => {
        return {
            root: {
                maxWidth: itemMaxWidth || '100%',
                minWidth: undefined,
                justifyContent: justify ?? 'left',
                borderRadius: borderRadius ?? undefined,
                lineHeight: itemHeight ?? undefined,
                height: itemHeight ?? undefined,
                fontSize: fontSize ?? undefined,
            },
            text: {
                minWidth: undefined,
            },
        };
    }, [itemMaxWidth, justify, borderRadius, itemHeight, fontSize]);

    const theme = React.useMemo(() => {
        try {
            return themeJSON ? createTheme(JSON.parse(themeJSON) as IPartialTheme) : undefined;
        } catch (ex) {
            /* istanbul ignore next */
            console.error('Cannot parse theme', ex);
        }
    }, [themeJSON]);

    const target = React.useRef<HTMLElement>(null);
    useResizeObserver(target, (entry) => {
        if (onResize) onResize(entry.contentRect.width, entry.contentRect.height);
    });

    return (
        <>
            <ThemeProvider theme={theme} applyTo="none" style={containerStyle} ref={target}>
                {items &&
                    items.map((item: TagListItem, index: number) => (
                        <Tag
                            key={item.id}
                            tabIndex={disabled ? -1 : tabIndex}
                            styles={tagStyles}
                            index={index}
                            item={item}
                            textColor={item.textColor}
                            borderColor={item.borderColor}
                            backgroundColor={item.backgroundColor}
                            iconName={item.iconName}
                            iconColor={item.iconColor}
                            disabled={disabled === true || item.enabled === false}
                            aria-setsize={items.length}
                            maxLength={300}
                        >
                            {item.name}
                        </Tag>
                    ))}
            </ThemeProvider>
        </>
    );
});
CanvasTagList.displayName = 'CanvasTagList';
