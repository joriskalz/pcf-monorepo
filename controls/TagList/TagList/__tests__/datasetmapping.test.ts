import { getTagItemsFromDataset } from '../components/DatasetMapping';
import { ItemColumns } from '../ManifestConstants';
import { MockDataSet, MockEntityRecord } from '../__mocks__/mock-datasets';

describe('DatasetMapping', () => {
    beforeEach(() => jest.clearAllMocks());

    it('returns correct props', () => {
        const items = [
            new MockEntityRecord('1', {
                [ItemColumns.Key]: 'commandOpen',
                [ItemColumns.DisplayName]: 'Open',
                [ItemColumns.IconName]: 'Open',
                [ItemColumns.IconColor]: 'blue',
                [ItemColumns.Enabled]: true,
                [ItemColumns.IconOnly]: false,
            }),
            new MockEntityRecord('2', {
                [ItemColumns.Key]: 'commandNew',
                [ItemColumns.DisplayName]: 'New',
                [ItemColumns.IconName]: 'New',
                [ItemColumns.IconColor]: 'blue',
                [ItemColumns.Enabled]: false,
                [ItemColumns.IconOnly]: true,
            }),
            new MockEntityRecord('3', {
                [ItemColumns.Key]: 'commandSave',
                [ItemColumns.DisplayName]: 'Save',
                [ItemColumns.IconName]: 'Save',
                [ItemColumns.IconColor]: 'green',
                [ItemColumns.Enabled]: false,
                [ItemColumns.IconOnly]: true,
            }),
        ];

        const tags = getTagItemsFromDataset(new MockDataSet(items));
        expect(tags).toMatchSnapshot();
    });

    it('removes duplicate keys', () => {
        const itemsWithDuplicates = [
            new MockEntityRecord('1', {
                [ItemColumns.Key]: 'key1',
                [ItemColumns.DisplayName]: 'Open',
            }),
            new MockEntityRecord('2', {
                [ItemColumns.Key]: 'key1',
                [ItemColumns.DisplayName]: 'Open',
            }),
            new MockEntityRecord('3', {
                [ItemColumns.Key]: 'key1',
                [ItemColumns.DisplayName]: 'Open',
            }),
            new MockEntityRecord('4', {
                [ItemColumns.Key]: '2',
                [ItemColumns.DisplayName]: 'Open',
            }),
            new MockEntityRecord('5', {
                [ItemColumns.Key]: '2',
                [ItemColumns.DisplayName]: 'Open',
            }),
            new MockEntityRecord('6', {
                [ItemColumns.Key]: 'unique3',
                [ItemColumns.DisplayName]: 'Open',
            }),
        ];

        const actions = getTagItemsFromDataset(new MockDataSet(itemsWithDuplicates));
        expect(actions.length).toBe(6);
        expect(actions[0].key).toBe('key1');
        expect(actions[1].key).toBe('key1_2');
        expect(actions[2].key).toBe('key1_3');
        expect(actions[3].key).toBe('2');
        expect(actions[4].key).toBe('2_2');
        expect(actions[5].key).toBe('unique3');
    });
});
