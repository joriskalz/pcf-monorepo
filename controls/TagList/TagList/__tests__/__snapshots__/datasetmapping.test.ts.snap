// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatasetMapping returns correct props 1`] = `
Array [
  Object {
    "backgroundColor": undefined,
    "borderColor": undefined,
    "data": MockEntityRecord {
      "getNamedReference": [MockFunction],
      "id": "1",
      "values": Object {
        "ItemDisplayName": "Open",
        "ItemEnabled": true,
        "ItemIconColor": "blue",
        "ItemIconName": "Open",
        "ItemIconOnly": false,
        "ItemKey": "commandOpen",
      },
    },
    "enabled": true,
    "iconColor": "blue",
    "iconName": "Open",
    "iconOnly": false,
    "id": "1",
    "key": "commandOpen",
    "name": "Open",
    "textColor": undefined,
    "visible": undefined,
  },
  Object {
    "backgroundColor": undefined,
    "borderColor": undefined,
    "data": MockEntityRecord {
      "getNamedReference": [MockFunction],
      "id": "2",
      "values": Object {
        "ItemDisplayName": "New",
        "ItemEnabled": false,
        "ItemIconColor": "blue",
        "ItemIconName": "New",
        "ItemIconOnly": true,
        "ItemKey": "commandNew",
      },
    },
    "enabled": false,
    "iconColor": "blue",
    "iconName": "New",
    "iconOnly": true,
    "id": "2",
    "key": "commandNew",
    "name": "New",
    "textColor": undefined,
    "visible": undefined,
  },
  Object {
    "backgroundColor": undefined,
    "borderColor": undefined,
    "data": MockEntityRecord {
      "getNamedReference": [MockFunction],
      "id": "3",
      "values": Object {
        "ItemDisplayName": "Save",
        "ItemEnabled": false,
        "ItemIconColor": "green",
        "ItemIconName": "Save",
        "ItemIconOnly": true,
        "ItemKey": "commandSave",
      },
    },
    "enabled": false,
    "iconColor": "green",
    "iconName": "Save",
    "iconOnly": true,
    "id": "3",
    "key": "commandSave",
    "name": "Save",
    "textColor": undefined,
    "visible": undefined,
  },
]
`;
